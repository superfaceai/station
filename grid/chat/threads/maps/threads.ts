/* eslint-disable jest/no-export */

import { RecordingProcessOptions, SuperfaceTest } from '@superfaceai/testing';

import { nockConfig } from '../../../test-config';

export const getThreadsTest = (
  provider: string,
  workspaces: [valid: string, invalid: string],
  options?: RecordingProcessOptions
): void => {
  describe(`chat/threads/${provider}`, () => {
    let superface: SuperfaceTest;

    describe('GetThreads', () => {
      beforeAll(() => {
        superface = new SuperfaceTest(
          {
            profile: 'chat/threads',
            provider,
            useCase: 'GetThreads',
          },
          nockConfig
        );
      });

      it('performs correctly', async () => {
        await expect(
          superface.run(
            {
              input: {
                workspace: workspaces[0],
              },
            },
            options
          )
        ).resolves.toMatchSnapshot();
      });

      it('fails', async () => {
        await expect(
          superface.run(
            {
              input: {
                workspace: workspaces[1],
              },
            },
            options
          )
        ).resolves.toMatchSnapshot();
      });
    });
  });
};
