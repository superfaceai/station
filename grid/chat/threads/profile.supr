"""
Chat Threads
Retrieve chat threads in the workspace.
"""
name = "chat/threads"
version = "1.0.0"

"""
Get Threads
Retrieve list of threads from the whole workspace.
"""
usecase GetThreads {
  input  {
    """
    Workspace
    Identifier of a workspace, for example a Discord guild or a Slack workspace.
    """
    workspace! string!
  }

  result {
    """
    Threads
    List of retrieved threads
    """
    threads! [Thread]

    """
    Rate Limit
    Rate limit details.
    """
    rateLimit RateLimit
  }

  error Error

  example Successful {
    input {
      workspace = 'workspace_id'
    }

    result {
      threads = [
        {
          id = '1'
          createdAt = 1546167104000
          channel = 'random'
        },
        {
          id = '2'
          createdAt = 1546167105000
          channel = 'random'
        }
      ]
    }
  }

  example Failed {
    input {
      workspace = 'not-existing-workspace'
    }

    error {
      title = 'workspace_not_found'
    }
  }
}

model Thread {
  """
  Id
  Unique identifier of the thread
  """
  id! string!

  """
  Created At
  UNIX timestamp (in milliseconds, e.g. 1095379198750) when the thread was created
  """
  createdAt! number!

  """
  Channel
  Unique identifier of the channel where the thread is located
  """
  channel! string!

  """
  Archive At
  UNIX timestamp (in milliseconds, e.g. 1095379198750) when the thread will be archived
  Some providers limit the lifetime of threads and automatically archive them after some time period of inactivity.
  """
  archiveAt number!

  """
  Name
  Name of retrieved thread
  """
  name string

  """
  Members Count
  Number of users participating in the thread
  """
  membersCount number

  """
  Messages Count
  Number of messages in the retrieved thread
  """
  messagesCount number

  """
  Last Message At
  UNIX timestamp (in milliseconds, e.g. 1095379198750) when was the last message sent to thread.
  """
  lastMessageAt number
}   

model Error {
  """
  Title
  A short, human-readable summary of the problem type.
  """
  title!

  """
  Detail
  A human-readable explanation specific to this occurrence of the problem.
  """
  detail

  """
  Rate Limit
  Rate limit details.
  """
  rateLimit RateLimit
}

model RateLimit {
  """
  Bucket
  Different parts of API may have different rate limits.
  Bucket identifies to which part of API the rate limits apply.
  """
  bucket string

  """
  Total Requests
  Total requests available in the time window.
  """
  totalRequests number

  """
  Remaining Requests
  Remaining requests available in the time window.
  """
  remainingRequests number

  """
  Remaining Requests available in percents
  Remaining requests available in the time window in percents.
  """
  remainingRequestsPercentage number

  """
  Reset Timestamp
  Timestamp when the rate limits will reset (in Unix Timestamp ms format).
  """
  resetTimestamp number

  """
  Reset After
  Number of seconds until reset of rate limit
  """
  resetAfter number

  """
  Retry After
  Number of seconds until user can retry the request
  """
  retryAfter number
}
