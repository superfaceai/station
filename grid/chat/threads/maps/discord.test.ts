import { getThreadsTest } from './threads';

getThreadsTest('discord', ['935962220104396881', '000000000000000000']);
