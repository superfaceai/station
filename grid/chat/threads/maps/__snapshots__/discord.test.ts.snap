// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`chat/threads/discord GetThreads fails 1`] = `
Err {
  "error": "HTTPError: Expected HTTP error
Properties: {
  \\"title\\": \\"Not found\\",
  \\"detail\\": \\"Unknown Guild\\"
}
AST Path: definitions[0].statements[0].responseHandlers[1].statements[1]
Original Map Location: Line 34, column 7",
}
`;

exports[`chat/threads/discord GetThreads performs correctly 1`] = `
Ok {
  "value": Object {
    "threads": Array [
      Object {
        "archiveAt": 1648300823821,
        "channel": "938614740995960844",
        "createdAt": 1643910369618,
        "id": "938779397664673793",
        "lastMessageAt": 1648214423830,
        "membersCount": 2,
        "messagesCount": 3,
        "name": "test thread",
      },
      Object {
        "archiveAt": 1648121147206,
        "channel": "938614740995960844",
        "createdAt": 1644399963026,
        "id": "940583259262308362",
        "lastMessageAt": 1648214397891,
        "membersCount": 1,
        "messagesCount": 7,
        "name": "new thread",
      },
    ],
  },
}
`;
